src/ctap/mod.rs

```diff
@@ -619,92 +756,74 @@ where
         }
 
         let rp_id_hash = Sha256::hash(rp_id.as_bytes());
-        let mut decrypted_credential = None;
-        let credentials = if let Some(allow_list) = allow_list {
-            let mut found_credentials = vec![];
-            for allowed_credential in allow_list {
-                match self.persistent_store.find_credential(
-                    &rp_id,
-                    &allowed_credential.key_id,
-                    !has_uv,
-                )? {
-                    Some(credential) => {
-                        found_credentials.push(credential);
-                    }
-                    None => {
-                        if decrypted_credential.is_none() {
-                            decrypted_credential = self.decrypt_credential_source(
-                                allowed_credential.key_id,
-                                &rp_id_hash,
-                            )?;
-                        }
-                    }
-                }
+        let mut credentials = if let Some(allow_list) = allow_list {
+            if let Some(credential) =
+                self.get_any_credential_from_allow_list(allow_list, &rp_id, &rp_id_hash, has_uv)?
+            {
+                vec![credential]
+            } else {
+                vec![]
             }
-            found_credentials
         } else {
-            // TODO(kaczmarczyck) use GetNextAssertion
             self.persistent_store.filter_credential(&rp_id, !has_uv)?
         };
+        // Remove user identifiable information without uv.
+        if !has_uv {
+            for credential in &mut credentials {
+                credential.other_ui = None;
+            }
+        }
+        credentials.sort_unstable_by_key(|c| c.creation_order);
 
-        let credential = if let Some(credential) = credentials.first() {
-            credential
-        } else {
-            decrypted_credential
-                .as_ref()
-                .ok_or(Ctap2StatusCode::CTAP2_ERR_NO_CREDENTIALS)?
-        };
-
+        // This check comes before CTAP2_ERR_NO_CREDENTIALS in CTAP 2.0.
+        // For CTAP 2.1, it was moved to a later protocol step.
         if options.up {
             (self.check_user_presence)(cid)?;
         }
 
-        self.increment_global_signature_counter()?;
+        let (credential, remaining_credentials) = credentials
+            .split_last()
```

Comment:
This seems costly. You convert back the remaining credentials to a `Vec` later. Why not use `pop` here?

Reply:
Leftover from an old revision. Now using `pop`, and also moving the credential later to avoid a few `clone`s later.